README.md

# Mojolicious-Plugin-Status [![](https://github.com/mojolicious/mojo-status/workflows/linux/badge.svg)](https://github.com/mojolicious/mojo-status/actions)

![Screenshot](https://raw.github.com/mojolicious/mojo-status/master/examples/status.png?raw=true)

  A server status ui for the [Mojolicious](https://mojolicious.org) real-time web framework. Note that this module is
  **EXPERIMENTAL** and should therefore only be used for debugging purposes.

```perl
use Mojolicious::Lite;

plugin 'Status';

app->start;
```

## Installation

  All you need is a one-liner, it takes less than a minute.

    $ curl -L https://cpanmin.us | perl - -M https://cpan.metacpan.org -n Mojolicious::Plugin::Status

  We recommend the use of a [Perlbrew](http://perlbrew.pl) environment.

## Want to know more?

  Take a look at our excellent
  [documentation](https://mojolicious.org/perldoc/Mojolicious/Plugin/Status)!


Changes

1.18  2021-03-24

1.17  2021-03-23
  - Fixed Mojolicious 9.11 compatibility.

1.16  2021-03-18
  - Updated ui with more consistent Mojolicious design.

1.15  2021-02-20
  - Improved Mojolicious::Plugin::Status to log if server is not supported.
  - Improved template design.

1.14  2021-02-09
  - Switched back to Sereal for serialization in Mojo::MemoryMap to reduce memory usage.

1.13  2020-12-10
  - Updated examples to use signatures.

1.12  2020-06-01
  - Updated project metadata.

1.11  2020-05-30
  - Improved .perltidyrc with more modern settings.

1.10  2020-05-26
  - Improved Mojo::MemoryMap to not require a Perl with 64-bit integer support.

1.09  2020-05-25
  - Switched from Sereal to Cpanel::JSON::XS for serialization in Mojo::MemoryMap.

1.08  2020-02-16
  - Fixed support for *BSD operating systems.

1.07  2020-02-14
  - Updated the File::Temp requirement to version 0.2308 to avoid flock bugs that could make web apps hang forever.

1.06  2020-02-14
  - Fixed displaying RSS on Linux correctly.

1.05  2020-02-13
  - Added support for tracking different response types.
  - Improved template design.
  - Fixed display bugs for large request paths and query strings.

1.04  2020-02-13
  - Improved template design.

1.03  2020-02-12
  - Added support for retaining a list of the slowest requests since server start.
  - Added reusable Mojo::MemoryMap module.
  - Improved template design.

1.02  2020-02-10
  - Switched to File::Map as storage backend, so this plugin should be much more reliable and safer to use now.
    (kraih, haarg, leont)

1.01  2019-10-07
  - The IPC mechanism used by this module is unreliable, so production use is strongly discouraged.

1.0  2018-09-17
  - Fixed a problem where subprocesses would show up in the worker list.

0.03  2018-07-23
  - Fixed package metadata.

0.02  2018-07-23
  - Fixed POD formatting bugs.

0.01  2018-07-22
  - First release.


examples/mixed.pl
use Mojolicious::Lite -signatures;
use Mojo::EventEmitter;
use Mojo::IOLoop;

plugin 'Status';

helper events => sub { state $events = Mojo::EventEmitter->new };

get '/' => 'dashboard';

get '/slow' => sub ($c) {
  $c->inactivity_timeout(3600);

  Mojo::IOLoop->timer(6 => sub { $c->redirect_to('dashboard') });
};

get '/superslow' => sub ($c) {
  $c->inactivity_timeout(3600);

  Mojo::IOLoop->timer(31 => sub { $c->redirect_to('dashboard') });
};

get '/subprocess' => sub ($c) {
  $c->inactivity_timeout(3600);

  Mojo::IOLoop->subprocess(sub { sleep 3 }, sub { $c->redirect_to('dashboard') });
};

get '/chat';

websocket '/channel' => sub ($c) {
  $c->inactivity_timeout(3600);

  # Forward messages from the browser
  $c->on(message => sub ($c, $message) { $c->events->emit(mojochat => $message) });

  # Forward messages to the browser
  my $cb = $c->events->on(mojochat => sub ($events, $message) { $c->send($message) });
  $c->on(finish => sub ($c, $code, $reason) { $c->events->unsubscribe(mojochat => $cb) });
};

# Minimal single-process WebSocket chat application for browser testing
app->start;
__DATA__

@@ dashboard.html.ep
<%= link_to Chat       => 'chat' %>
<%= link_to Slow       => 'slow' %>
<%= link_to Superslow  => 'superslow' %>
<%= link_to Subprocess => 'subprocess' %>
<%= link_to Status     => 'mojo_status' %>

@@ chat.html.ep
<form onsubmit="sendChat(this.children[0]); return false"><input></form>
<div id="log"></div>
<script>
  var ws  = new WebSocket('<%= url_for('channel')->to_abs %>');
  ws.onmessage = function (e) {
    document.getElementById('log').innerHTML += '<p>' + e.data + '</p>';
  };
  function sendChat(input) { ws.send(input.value); input.value = '' }
</script>


lib/Mojo/MemoryMap.pm
package Mojo::MemoryMap;
use Mojo::Base -base;

use File::Map qw(map_anonymous);
use Mojo::File qw(tempfile);
use Mojo::MemoryMap::Writer;

sub new {
  my ($class, $size) = @_;
  my $self = $class->SUPER::new(size => $size // 52428800, usage => 0);

  $self->{tempfile} = tempfile->touch;
  map_anonymous my $map_a, $self->{size}, 'shared';
  map_anonymous my $map_b, $self->{size}, 'shared';
  map_anonymous my $current, 1, 'shared';
  substr $current, 0, 1, 'a';
  $self->{map} = {a => \$map_a, b => \$map_b, current => \$current};
  $self->writer->store({});

  return $self;
}

sub size { shift->{size} }

sub usage { shift->{usage} }

sub writer {
  my $self = shift;
  my $fh   = $self->{fh}{$$} ||= $self->{tempfile}->open('>');
  return Mojo::MemoryMap::Writer->new(fh => $fh, map => $self->{map}, usage => \$self->{usage});
}

1;

=encoding utf8

=head1 NAME

Mojo::MemoryMap - Safely use anonymous memory mapped segments

=head1 SYNOPSIS

  use Mojo::MemoryMap;

  my $map = Mojo::MemoryMap->new(4096);
  say $map->usage;
  $map->writer->store({foo => 123});
  say $map->writer->fetch->{foo};
  say $map->writer->change(sub { delete $_->{foo} });
  say $map->usage;

=head1 DESCRIPTION

L<Mojo::MemoryMap> uses L<File::Map> to allow you to safely cache mutable data structures in anonymous mapped memory
segments, and share it between multiple processes.

=head1 METHODS

L<Mojo::MemoryMap> inherits all methods from L<Mojo::Base> and implements the following new ones.

=head2 new

  my $map = Mojo::MemoryMap->new;
  my $map = Mojo::MemoryMap->new(4096);

Construct a new L<Mojo::MemoryMap> object, defaults to a L</"size"> of C<52428800> bytes (50 MiB).

=head2 size

  my $size = $map->size;

Size of anonymous memory segments in bytes (will use twice as much memory for safety).

=head2 usage

  my $usage = $map->usage;

Current usage of anonymous memory segment in bytes.

=head2 writer

  my $writer = $map->writer;

Acquire exclusive lock and return L<Mojo::MemoryMap::Writer> object. Allowing the shared data structure to be retrieved
and modified safely. The lock is released when the writer object is destroyed.

  # Retrieve data
  my $data = $map->writer->fetch;

  # Modify data safely
  my $writer = $map->writer;
  my $data = $writer->fetch;
  $data->{foo} += 23;
  $writer->store($data);
  undef $writer;

  # Modify data safely (with less code)
  $map->writer->change(sub { $_->{foo} += 23 });

=head1 SEE ALSO

L<Mojolicious::Plugin::Status>, L<Mojolicious::Guides>, L<https://mojolicious.org>.

=cut


lib/Mojo/MemoryMap/Writer.pm
package Mojo::MemoryMap::Writer;
use Mojo::Base -base;

use Fcntl qw(:flock);
use Sereal qw(decode_sereal encode_sereal);

sub DESTROY { flock(shift->{fh}, LOCK_UN) or die "Couldn't flock: $!" }

sub change {
  my ($self, $cb) = @_;
  my $data = $self->fetch;
  $cb->($_) for $data;
  return $self->store($data);
}

sub fetch {
  my $self = shift;
  my $len  = unpack 'N', substr(${$self->{map}}, 0, 4);
  return decode_sereal(substr(${$self->{map}}, 4, $len));
}

sub new {
  my $self = shift->SUPER::new(@_);
  flock($self->{fh}, LOCK_EX) or die "Couldn't flock: $!";
  return $self;
}

sub store {
  my ($self, $data) = @_;

  my $bin   = encode_sereal($data);
  my $bytes = pack('N', length $bin) . $bin;

  ${$self->{usage}} = my $usage = length $bytes;
  return undef if $usage > length ${$self->{map}};
  substr ${$self->{map}}, 0, $usage, $bytes;

  return 1;
}

1;

=encoding utf8

=head1 NAME

Mojo::MemoryMap::Writer - Writer

=head1 SYNOPSIS

  use Mojo::MemoryMap::Writer;

  my $writer = Mojo::MemoryMap::Writer->new(map => $map);
  $writer->store({foo => 123});
  say $writer->fetch->{foo};

=head1 DESCRIPTION

L<Mojo::MemoryMap::Writer> is a scope guard for L<Mojo::MemoryMap> that allows you to write safely to anonymous mapped
memory segments.

=head1 METHODS

L<Mojo::MemoryMap::Writer> inherits all methods from L<Mojo::Base> and implements the following new ones.

=head2 change

  my $bool = $writer->change(sub {...});

Fetch data, modify it in the closure and store it again right away.

  # Remove a value safely
  $writer->change(sub { delete $_->{foo} });

=head2 fetch

  my $data = $writer->fetch;

Fetch data.

=head2 new

  my $writer = Mojo::MemoryMap::Writer->new;
  my $writer = Mojo::MemoryMap::Writer->new(map => $map);
  my $writer = Mojo::MemoryMap::Writer->new({map => $map});

Construct a new L<Mojo::MemoryMap::Writer> object.

=head2 store

  my $bool = $writer->store({foo => 123});

Store data, replacing all existing data.

=head1 SEE ALSO

L<Mojolicious::Plugin::Status>, L<Mojolicious::Guides>, L<https://mojolicious.org>.

=cut


lib/Mojolicious/Plugin/Status.pm
package Mojolicious::Plugin::Status;
use Mojo::Base 'Mojolicious::Plugin';

use BSD::Resource qw(getrusage);
use Time::HiRes qw(time);
use Mojo::File qw(path);
use Mojo::IOLoop;
use Mojo::MemoryMap;
use Mojo::Util qw(humanize_bytes);

use constant MACOS => $^O eq 'darwin';

our $VERSION = '1.18';

sub register {
  my ($self, $app, $config) = @_;

  # Config
  my $prefix = $config->{route} // $app->routes->any('/mojo-status');
  $prefix->to(return_to => $config->{return_to} // '/');
  $self->{slowest} = $config->{slowest} // 10;

  # Initialize cache
  my $map = $self->{map} = Mojo::MemoryMap->new($config->{size});
  $map->writer->store({processed => 0, started => time, stats => _stats(), slowest => []});

  # Only the two built-in servers are supported for now
  $app->hook(before_server_start => sub { $self->_start(@_) });

  # Static files
  my $resources = path(__FILE__)->sibling('resources');
  push @{$app->static->paths}, $resources->child('public')->to_string;

  # Templates
  push @{$app->renderer->paths}, $resources->child('templates')->to_string;

  # Routes
  $prefix->get('/' => [format => ['json']] => {format => undef, mojo_status => $self} => \&_dashboard)
    ->name('mojo_status');
}

sub _activity {
  my $all = shift;

  # Workers
  my @table;
  for my $pid (sort keys %{$all->{workers}}) {
    my $worker = $all->{workers}{$pid};
    my $cpu    = sprintf '%.2f', $worker->{utime} + $worker->{stime};
    my @worker = ($pid, $cpu, humanize_bytes($worker->{maxrss}));

    # Connections
    my $connections = $worker->{connections};
    if (keys %$connections) {
      my $repeat;
      for my $cid (sort keys %$connections) {
        my $conn = $connections->{$cid};
        @worker = ('', '', '') if $repeat++;
        my $bytes_read    = humanize_bytes $conn->{bytes_read};
        my $bytes_written = humanize_bytes $conn->{bytes_written};
        my $rw            = "$bytes_read/$bytes_written";
        my @conn          = ($conn->{client}, $rw, $conn->{processed});

        # Request
        if (my $req = $conn->{request}) {
          my $active = $req->{finished} ? 0 : 1;
          my ($rid, $proto) = @{$req}{qw(request_id protocol)};

          my $str = "$req->{method} $req->{path}";
          $str .= "?$req->{query}"    if $req->{query};
          $str .= " → $req->{status}" if $req->{status};

          my $finished = $active ? time : $req->{finished};
          my $time     = sprintf '%.2f', $finished - $req->{started};
          push @table, [@worker, @conn, $rid, $active, $time, $proto, $str];
        }
        else { push @table, [@worker, @conn] }
      }
    }
    else { push @table, \@worker }
  }

  return \@table;
}

sub _dashboard {
  my $c = shift;

  my $map = $c->stash('mojo_status')->{map};
  if ($c->param('reset')) {
    $map->writer->change(sub { @{$_}{qw(slowest stats)} = ([], _stats()) });
    return $c->redirect_to('mojo_status');
  }

  my $all = $map->writer->fetch;
  $c->respond_to(
    html => sub {
      $c->render(
        'mojo-status/dashboard',
        now      => time,
        usage    => humanize_bytes($map->usage),
        size     => humanize_bytes($map->size),
        activity => _activity($all),
        slowest  => _slowest($all),
        all      => $all
      );
    },
    json => {json => $all}
  );
}

sub _read_write {
  my ($all, $id) = @_;
  return unless my $stream = Mojo::IOLoop->stream($id);
  @{$all->{workers}{$$}{connections}{$id}}{qw(bytes_read bytes_written)}
    = ($stream->bytes_read, $stream->bytes_written);
}

sub _request {
  my ($self, $c) = @_;

  # Request start
  my $tx    = $c->tx;
  my $id    = $tx->connection;
  my $req   = $tx->req;
  my $url   = $req->url->to_abs;
  my $proto = $tx->is_websocket ? 'ws' : 'http';
  $proto .= 's' if $req->is_secure;
  $self->{map}->writer->change(sub {
    $_->{workers}{$$}{connections}{$id}{request} = {
      request_id => $req->request_id,
      method     => $req->method,
      protocol   => $proto,
      path       => $url->path->to_abs_string,
      query      => $url->query->to_string,
      started    => time
    };
    _read_write($_, $id);
    $_->{workers}{$$}{connections}{$id}{client} = $tx->remote_address;
  });

  # Request end
  $tx->on(
    finish => sub {
      my $tx = shift;
      $self->{map}->writer->change(sub {
        return unless my $worker = $_->{workers}{$$};

        my $code = $tx->res->code || 0;
        if    ($code > 499) { $_->{stats}{server_error}++ }
        elsif ($code > 399) { $_->{stats}{client_error}++ }
        elsif ($code > 299) { $_->{stats}{redirect}++ }
        elsif ($code > 199) { $_->{stats}{success}++ }
        elsif ($code)       { $_->{stats}{info}++ }

        @{$worker->{connections}{$id}{request}}{qw(finished status)} = (time, $code);
        $worker->{connections}{$id}{processed}++;
        $worker->{processed}++;
        $_->{processed}++;
      });
    }
  );
}

sub _rendered {
  my ($self, $c) = @_;

  my $id = $c->tx->connection;
  $self->{map}->writer->change(sub {
    return unless my $conn = $_->{workers}{$$}{connections}{$id};
    return unless my $req  = $conn->{request};
    $req->{time} = time - $req->{started};
    @{$req}{qw(client status worker)} = ($conn->{client}, $c->res->code, $$);

    my $slowest = $_->{slowest};
    @$slowest = sort { $b->{time} <=> $a->{time} } @$slowest, $req;
    my %seen;
    @$slowest = grep { !$seen{"$_->{method} $_->{path}"}++ } @$slowest;
    pop @$slowest while @$slowest > $self->{slowest};
  });
}

sub _resources {
  my $self = shift;

  $self->{map}->writer->change(sub {
    @{$_->{workers}{$$}}{qw(utime stime maxrss)} = (getrusage)[0, 1, 2];

    # macOS actually returns bytes instead of kilobytes
    $_->{workers}{$$}{maxrss} = $_->{workers}{$$}{maxrss} * 1000 unless MACOS;

    for my $id (keys %{$_->{workers}{$$}{connections}}) { _read_write($_, $id) }
  });
}

sub _slowest {
  my $all = shift;

  my @table;
  for my $req (@{$all->{slowest}}) {
    my $str = "$req->{method} $req->{path}";
    $str .= "?$req->{query}"    if $req->{query};
    $str .= " → $req->{status}" if $req->{status};
    my $time = sprintf '%.2f', $req->{time};
    push @table, [$time, $str, @{$req}{qw(request_id worker client started)}];
  }

  return \@table;
}

sub _start {
  my ($self, $server, $app) = @_;

  return $app->log->warn('Server not suported by Mojolicious::Plugin::Status')
    unless $server->isa('Mojo::Server::Daemon');

  # Register started workers
  Mojo::IOLoop->next_tick(sub {
    $self->{map}->writer->change(sub { $_->{workers}{$$} = {started => time, processed => 0} });
  });

  # Remove stopped workers
  $server->on(
    reap => sub {
      my ($server, $pid) = @_;
      $self->{map}->writer->change(sub { delete $_->{workers}{$pid} });
    }
  ) if $server->isa('Mojo::Server::Prefork');

  # Collect stats
  $app->hook(after_build_tx  => sub { $self->_tx(@_) });
  $app->hook(before_dispatch => sub { $self->_request(@_) });
  $app->hook(after_dispatch  => sub { $self->_rendered(@_) });
  Mojo::IOLoop->next_tick(sub { $self->_resources });
  Mojo::IOLoop->recurring(5 => sub { $self->_resources });
}

sub _stats { {started => time, info => 0, success => 0, redirect => 0, client_error => 0, server_error => 0} }

sub _stream {
  my ($self, $id) = @_;

  my $stream = Mojo::IOLoop->stream($id);
  $stream->on(
    close => sub {
      $self->{map}->writer->change(sub { delete $_->{workers}{$$}{connections}{$id} if $_->{workers}{$$} });
    }
  );
}

sub _tx {
  my ($self, $tx, $app) = @_;

  $tx->on(
    connection => sub {
      my ($tx, $id) = @_;

      my $map = $self->{map};
      return if $map->writer->fetch->{workers}{$$}{connections}{$id};

      $map->writer->change(sub {
        $_->{workers}{$$}{connections}{$id} = {started => time, processed => 0, bytes_read => 0, bytes_written => 0};
      });
      $self->_stream($id);
    }
  );
}

1;

=encoding utf8

=head1 NAME

Mojolicious::Plugin::Status - Mojolicious server status

=head1 SYNOPSIS

  # Mojolicious
  $self->plugin('Status');

  # Mojolicious::Lite
  plugin 'Status';

  # Secure access to the server status ui with Basic authentication
  my $under = $self->routes->under('/status' => sub ($c) {
    return 1 if $c->req->url->to_abs->userinfo eq 'Bender:rocks';
    $c->res->headers->www_authenticate('Basic');
    $c->render(text => 'Authentication required!', status => 401);
    return undef;
  });
  $self->plugin('Status' => {route => $under});

=head1 DESCRIPTION

=begin html

<p>
  <img alt="Screenshot" src="https://raw.github.com/mojolicious/mojo-status/master/examples/status.png?raw=true"
    width="600px">
</p>

=end html

L<Mojolicious::Plugin::Status> is a L<Mojolicious> plugin providing a server status ui for L<Mojo::Server::Daemon> and
L<Mojo::Server::Prefork>. Note that this module is B<EXPERIMENTAL> and should therefore only be used for debugging
purposes.

=head1 OPTIONS

L<Mojolicious::Plugin::Status> supports the following options.

=head2 return_to

  # Mojolicious::Lite
  plugin Status => {return_to => 'some_route'};

Name of route or path to return to when leaving the server status ui, defaults to C</>.

=head2 route

  # Mojolicious::Lite
  plugin Status => {route => app->routes->any('/status')};

L<Mojolicious::Routes::Route> object to attach the server status ui to, defaults to generating a new one with the
prefix C</mojo-status>.

=head2 size

  # Mojolicious::Lite
  plugin Status => {size => 1234};

Size of anonymous mapped memory to use for storing statistics, defaults to C<52428800> (50 MiB).

=head2 slowest

  # Mojolicious::Lite
  plugin Status => {slowest => 5}; 

Number of slowest requests to track, defaults to C<10>.

=head1 METHODS

L<Mojolicious::Plugin::Status> inherits all methods from L<Mojolicious::Plugin> and implements the following new ones.

=head2 register

  my $route = $plugin->register(Mojolicious->new);

Register renderer and helper in L<Mojolicious> application.

=head1 BUNDLED FILES

The L<Mojolicious::Plugin::Status> distribution includes a few files with different licenses that have been bundled for
internal use.

=head2 Artwork

  Copyright (C) 2018, Sebastian Riedel.

Licensed under the CC-SA License, Version 4.0 L<http://creativecommons.org/licenses/by-sa/4.0>.

=head2 Bootstrap

  Copyright (C) 2011-2018 The Bootstrap Authors.

Licensed under the MIT License, L<http://creativecommons.org/licenses/MIT>.

=head2 Font Awesome

  Copyright (C) Dave Gandy.

Licensed under the MIT License, L<http://creativecommons.org/licenses/MIT>, and the SIL OFL 1.1,
L<http://scripts.sil.org/OFL>.

=head1 AUTHOR

Sebastian Riedel, C<sri@cpan.org>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2018-2021, Sebastian Riedel and others.

This program is free software, you can redistribute it and/or modify it under the terms of the Artistic License version
2.0.

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<https://mojolicious.org>.

=cut


Makefile.PL
use 5.016;

use strict;
use warnings;

use ExtUtils::MakeMaker;

# File::Temp 0.2308+ is required to avoid flock bugs
WriteMakefile(
  NAME         => 'Mojolicious::Plugin::Status',
  VERSION_FROM => 'lib/Mojolicious/Plugin/Status.pm',
  ABSTRACT     => 'Mojolicious server status',
  AUTHOR       => 'Sebastian Riedel <sri@cpan.org>',
  LICENSE      => 'artistic_2',
  META_MERGE   => {
    dynamic_config => 0,
    'meta-spec'    => {version   => 2},
    no_index       => {directory => ['t']},
    prereqs        => {runtime   => {requires => {perl => '5.016'}}},
    resources      => {
      bugtracker => {web => 'https://github.com/mojolicious/mojo-status/issues'},
      homepage   => 'https://mojolicious.org',
      license    => ['http://www.opensource.org/licenses/artistic-license-2.0'],
      repository => {
        type => 'git',
        url  => 'https://github.com/mojolicious/mojo-status.git',
        web  => 'https://github.com/mojolicious/mojo-status',
      },
      x_IRC => {url => 'irc://irc.freenode.net/#mojo', web => 'https://webchat.freenode.net/#mojo'}
    },
  },
  PREREQ_PM => {Mojolicious => '9.11', 'BSD::Resource' => 0, 'Sereal' => 0, 'File::Map' => 0, 'File::Temp' => '0.2308'},
  test      => {TESTS       => 't/*.t'}
);
